File: lib/aura/kernel/hooks.rb
# frozen_string_literal: true

module Aura
  module Kernel
    class Hooks
      def initialize
        @hooks = {}
      end

      def register(name, &block)
        @hooks[name] ||= []
        @hooks[name] << block
      end

      def unregister(name, hook_proc)
        @hooks[name]&.delete(hook_proc)
      end

      # Runs hooks for a given name.
      # If any hook returns explicitly false (not nil), the operation may be considered halted/rejected.
      # Returns true if all hooks pass (or no hooks exist), false if any hook returns false.
      def run(name, *args)
        return true unless @hooks[name]

        @hooks[name].each do |hook|
          result = hook.call(*args)
          return false if result == false
        end
        true
      end
    end
  end
end

File: src/core/kernel/hooks.ts
import type { HookFn, IHooks } from './interfaces.js';

export class Hooks implements IHooks {
  private hookMap: Record<string, HookFn[]> = {};

  public register(name: string, hookFn: HookFn): void {
    if (!this.hookMap[name]) {
      this.hookMap[name] = [];
    }
    this.hookMap[name].push(hookFn);
  }

  public unregister(name: string, hookFn: HookFn): void {
    if (!this.hookMap[name]) return;
    const idx = this.hookMap[name].indexOf(hookFn);
    if (idx !== -1) {
      this.hookMap[name].splice(idx, 1);
    }
  }

  /**
   * Runs all hooks registered under `name`.
   * Returns false if any synchronous hook explicitly returned false (blocking).
   * Async hooks are fire-and-forget from the perspective of the return value.
   */
  public run(name: string, ...args: unknown[]): boolean {
    const list = this.hookMap[name];
    if (!list) return true;

    for (const hook of list) {
      const result = hook(...args);
      if (result === false) {
        return false;
      }
    }
    return true;
  }
}
